addons/sale_timesheet/tests/test_reinvoice.py

```diff
@@ -366,3 +366,44 @@ def test_reversed_invoice_reinvoice_with_period(self):
 
         # The actual test :
         wizard.create_invoices()  # No exception should be raised, there is indeed something to be invoiced since it was reversed
+
+    def test_project_update_reinvoiced_product(self):
+        project_product, expense_product = self.env['product.product'].create([{
+            'name': 'Project creation',
+            'type': 'service',
+            'service_tracking': 'task_in_project',
+        }, {
+            'name': 'Expense Product',
+            'expense_policy': 'sales_price',
+            'list_price': 20,
+        }])
+        self.env['sale.order.line'].create({
+            'product_id': project_product.id,
+            'order_id': self.sale_order.id,
+        })
+        self.sale_order.action_confirm()
+        project = self.sale_order.project_ids
+        self.assertTrue(project, 'Project should have been created on sale order confirmation')
+
+        vendor_bill = self.env['account.move'].create({
+            'partner_id': self.partner_a.id,
+            'invoice_date': self.sale_order.date_order,
+            'journal_id': self.env['account.journal'].search([('code', '=', 'BILL'), ('company_id', '=', self.env.company.id)]).id,
+            'move_type': 'in_invoice',
+        })
+        self.env['account.move.line'].create({
+            'product_id': expense_product.id,
+            'move_id': vendor_bill.id,
+            'account_id': self.env['account.account'].search([('code', '=', '600000'), ('company_id', '=', self.env.company.id)]).id,
+            'analytic_distribution': {project.analytic_account_id.id: 100},
+            'price_unit': 20,
+        })
+        vendor_bill.action_post()
+        self.assertEqual(project.analytic_account_id.vendor_bill_count, 1, 'Vendor bill should be liked to project account')
+
+        # An analytic line is created for the vendor bill, and another for the sale order line (created by the re-invoiced expense product)
+        # Only the original vendor bill amount should appear on the project update, to stay consistent with the corresponding hr_expense behavior
+        updates = project._get_profitability_items()
+        self.assertEqual(vendor_bill.line_ids.analytic_line_ids.category, 'vendor_bill')
+        self.assertEqual(self.sale_order.order_line.analytic_line_ids.category, 'vendor_bill')
+        self.assertEqual(updates['costs']['total']['billed'], -20, 'Only the vendor bill should be deducted')
```

Comment:
I don't think it's worth bothering checking the category of the aal generated. What matters to us is the result of the updates. I'd also suggest to check the section 'vendor bills' in addition of the total costs. This will ensure that the 'expense' is computed into the correct section. Otherwise, lgtm. Glad to see that the category was indeed enough and did not break other use case :smile:

Reply:
@damrOdoo Well, I've just had a closer look at that and it's actually not in the 'vendor bill' category (Despite the analytic line having that category), like everything else that goes through `_get_costs_items_from_purchase` it's in `other_purchase_costs` https://github.com/odoo/odoo/blob/17.0/addons/project_account/models/project_project.py#L72-L80.
But I guess that doesn't really matter since the project update links to the vendor bill anyway.

Also I'm not sure how I missed this before but we were wrong on one point: there is only 1 analytic line that is linked to both the vendor bill and the sol, so this workaround only works because of that quirk. This all kinda sucks but so did the other alternatives we came up with on Thursday so maybe it's still the better option... What do you think ?

Reply:
For the first part of the comment, your misunderstanding is due to bad wording on our part. There is actually 2 sections in project profitability  with vendor bills. One named ' Purchases', which contains the purchase orders linked to a sale order, and one named 'Vendor bills' which contains purchase orders where the user manually added the analytic account on it. Here's where we f*cked up : we named the first section 'vendor bills', and the second section 'other purchase cost' inside the dict of values that we pass through the 'get_profitability method' The issue is that the second section was created later than the first, and we did not think of updating the names to be more fitting.

About the second part, it does not change much in our work flow. If that is how the sales want to handle those analytic lines, we can't do much about it, and the work around we use in our fix is still valid. And this solution has the advantage of being valid both for new entries and for client which already did the use case and have this data structure in their db.